### ISSUE TITLE
Add a tutorial scenario

### ISSUE THREAD

**Issue by CelticMinstrel** [calref/cboe#174]
The game could offer to automatically place newly-created parties in this scenario.

The tutorial scenario could also double as a way to increase the party's level if they wish to enter a higher-level scenario.

**Comment by NQNStudios** [comment]
I think, new parties could be prompted to play the tutorial, but also--

I would like to add a "Tutorial" button to the main menu, which puts the player in the tutorial scenario with a *default party*. Because BoE character creation is wildly intimidating by modern RPG standards, even I don't wanna do it sometimes.

I had to dig very deep to find this, but it's an old blog post from JV about how front-loaded choices are overwhelming: https://jeff-vogel.blogspot.com/2011/08/dont-ask-questions-until-player-can.html

**Comment by CelticMinstrel** [comment]
That's not a bad idea. The tutorial scenario could even be built to use the hypothetical "scenario prefab party" feature that I wanted to add to the game.

There are some scenarios that come with a saved game file with a special party that the designer recommends using – it would just be a way of making that officially acknowledged by the engine. (Of course, it would not force you to use the prebuilt party, though an exception could be made for the tutorial.)

I thought there was an issue about that but I can't seem to find it, so maybe I forgot to create it…

**Comment by NQNStudios** [comment]
> There are some scenarios that come with a saved game file with a special party that the designer recommends using – it would just be a way of making that officially acknowledged by the engine. (Of course, it would not force you to use the prebuilt party, though an exception could be made for the tutorial.)

Oh, I've been thinking about this too. Prefabs absolutely should be supported in-engine. And I don't think there's an issue for it.

**Comment by NQNStudios** [comment]
When I'm demoing cboe at the local playtesting event, I'll be telling everyone to start with the tutorial scenario. The first person who plays through it will get all the help dialogs the first time they do things (which is good for the tutorial) but the 2nd person to play won't get any. So I propose that when starting the tutorial scenario, if the preferences are already storing which dialogs have already been seen, these values should be temporarily wiped.

Then after the tutorial ends or is quit, the "help received" preference should be set to the OR'd value of the original values and the values shown during the tutorial.

**Comment by CelticMinstrel** [comment]
Is manually resetting the help after each playtester not an option?

**Comment by NQNStudios** [comment]
Not a very good one 👿 I will forget to do that, 100% every time.

This also would be nice for the case where you're showing a friend BoE on your own computer.

**Comment by NQNStudios** [comment]
Ok-- the playtesting event has limited space and they didn't select Open Blades of Exile to be in it. That complicated feature seems less important now.

Here's something simpler -- starting the tutorial with help dialogs stored as already shown, we could bring up the yes/no prompt and suggest resetting them if it's a new player.

**Comment by NQNStudios** [comment]
@CelticMinstrel Do you have any interest in designing the tutorial?

**Comment by CelticMinstrel** [comment]
I don't mind doing all the work of implementing it, but I'd prefer not to be completely responsible for deciding "what does it need to do".

**Comment by NQNStudios** [comment]
We could make a Google Doc or other type of shared/synchronized document for outlining/designing it?

**Comment by CelticMinstrel** [comment]
That's one option.

**Comment by NQNStudios** [comment]
It could be organized as a hallway with stairs/portals for different groups of sub-tutorials, and you can just walk past ones you already know, and right out the scenario. So the intro message is something like, "To move, click in a direction or use the arrow keys. If you've never played a computer RPG, move north to the first portal."

Actually now that I think of it I might be down to do the implementation.

**Comment by NQNStudios** [comment]
Separating things non-linearly would also make it modular so not every feature has to be explained for the tutorial to be released in a build. We could add more later without breaking anything.

**Comment by CelticMinstrel** [comment]
> It could be organized as a hallway with stairs/portals for different groups of sub-tutorials, and you can just walk past ones you already know, and right out the scenario.

This sounds _very_ dubious to me. I do agree that it makes sense to have an option to skip over the tutorial, whether it be a secret exit like in Blades of Avernum or a "Do you want to play the tutorial?" question to decide whether you even go into the tutorial at all.

> So the intro message is something like, "To move, click in a direction or use the arrow keys. If you've never played a computer RPG, move north to the first portal."

This part makes sense. Seems like a good start

> Actually now that I think of it I might be down to do the implementation.

That's fine too. It could also be shared.

> Separating things non-linearly would also make it modular so not every feature has to be explained for the tutorial to be released in a build. We could add more later without breaking anything.

Though I kinda see what you mean, a non-linear tutorial just strikes me as strange. I think it should probably be linear. That said, I'm open to hearing more detailed suggestions either way.

**Comment by NQNStudios** [comment]
Non-linear was not the right phrase. I think I mean *modular*--like each thing being taught should be, in the implementation logic and in map placement, separate from the other parts. So going in and adding or replacing sections should be safe--instead of things stuffed into a map where space is limited, we set it up where there's plenty of room to work and insert in the middle. i.e. each tutorial section is a town, we don't mix sections in the same town and set ourselves up for future tetris problems.

The "lessons" would still certainly be ordered so they build from basic to complex and a brand-new player would go through each one.

**Comment by CelticMinstrel** [comment]
Okay, yeah, making it modular like that doesn't sound bad. We'd probably want to use the smallest town size for most of them, but some could potentially be larger if it makes sense.

(Hmm, should we add a way to change the size of a town after it's created…?)

**Comment by NQNStudios** [comment]
@CelticMinstrel Here's an outline I made of things that could be in the tutorial. If you agree with the basic idea, you could edit the issue description to move this checklist into it. Every section would be optional and modular.

I think that if there's one thing we can assume people already know, it's how to click on buttons and interact with a UI if things are clearly labeled. The tutorial only needs to address dialogs when they are complex or unintuitive (I think targeting a PC in the spell dialog is kinda weird, for example.)

These are not all in a sequential order yet in terms of which order a new player should see them in. Kinda, but kinda not.

- [ ] **Have you ever played an RPG?** Section for people who have no idea about any of it. (Important to me because I have family members who might want to check it out because they're interested in what I do.) Could be all in dialogs, explaining things like 'You have a party of characters who will get stronger as you make choices and explore a fantasy story.'
- [ ] **How to move** - Clicking or using directional keys (numpad/arrows) to move the party. Diagonal movement. NPCs move once when you move once. Don't step on lava. Special spots: go to them.
- [ ] **How to explore thoroughly using the map**
- [ ] **Exile Lore** - the player could walk through a little interactive cutscene summarizing Exile 1, 2, and 3. Explain all the playable species culture/lore, Exile and Valorim, the Empire, etc.
- [ ] Talking to NPCs. Using the 'Ask about' button.
- [ ] **Equipping/unequipping items, using items** - Fill the party's inventory with some stuff and have them equip it/use it before they can move on. Explain the basic potions and most common consumables. (Because this tutorial is modular, they *should not need* any of it for later parts. Maybe we wipe their inventory afterwards.)
- [ ] **Shopping**
- [ ] **How to cast spells (peace mode)**
- [ ] **Look, Use, Reading Signs** - Opening/closing doors. Finding items in containers. Stuff like that.
- [ ] **Secret doors and how to find them**
- [ ] **Lockpicking/bash doors and chests, traps**
- [ ] **Combat** - Party turn order. AP. Missile attacks. Combat spells. Wall spells. Basic tactical recommendations for the uninitiated.
- [ ] **Outdoor combat** - wandering monsters, outdoor combat arenas, fleeing battles
- [ ] **resting, town wait, eating food & hunger effects**
- [ ] **Status effects** - poison, disease, all the basics. Curing, cleansing.
- [ ] **Alchemy**
- [ ] **Flying**
- [ ] **Death and resurrection (the various spells, destoning, risk of dusting, vs. paying healers)**
- [ ] **Damage types and resistances** - damage vs. undead, fire damage vs. fire enemies, cold vs. undead, etc.
- [ ] **Spells requiring reagents**
- [ ] **Simulacrum**
- [ ] **Finding custom scenarios, asking for help, reporting bugs**
- [ ] **End of tutorial** - Could wipe the party's inventory/experience and ask whether to start VoDT

Alternately, I could make a new Tutorial issue so that both of us would have edit permissions on the description/checklist.

**Comment by CelticMinstrel** [comment]
Broadly speaking, I don't see anything on that list that I think shouldn't be tere.

The tutorial can probably make extensive use of the Display Picture node to give instructions – take a screenshot, slap text over it. If we do that we should include clean (text-free) versions of the image too, for future localization purposes (not that that's happening anytime soon, but my opinion is that it'd be more annoying to have to redo them later).

> Could be all in dialogs, explaining things like 'You have a party of characters who will get stronger as you make choices and explore a fantasy story.'

Would this be a good place to use the Story Dialog node? This is the concept of the Exile II introductory text.

> Because this tutorial is modular, they should not need any of it for later parts. Maybe we wipe their inventory afterwards.

I don't really like wiping their inventory. We can arrange for the items to be destroyed for Reasons, perhaps.


> Darkness, torches, light spells

No need to cover special light modes, right? Like dungeons that suck your light.

> Could wipe the party's inventory/experience and ask whether to start VoDT

Again, not a fan of wiping the inventory.

> If you agree with most of this, I could make a new Tutorial issue so that both of us would have edit permissions on the description/checklist.

Sure.

**Comment by NQNStudios** [comment]
Closed in favor of #705 which has an updated version of this checklist fixing some suggestions.
